File: 资料/图表/Matlab绘图代码/基础散点图、带趋势线的散点图、分组散点图、气泡图、基础折线图、多系列折线图、带误差线的折线图、阶梯折线图、基础柱状图、水平柱状图.m
% 10种基础图表的MATLAB实现
% 包含中文显示设置，代码可直接运行

% 设置中文字体支持，确保中文正常显示
try
    set(0, 'DefaultTextFontName', 'SimHei');   % 优先尝试SimHei字体
    set(0, 'DefaultAxesFontName', 'SimHei');
catch
    try
        set(0, 'DefaultTextFontName', 'WenQuanYi Micro Hei');  % 次选文泉驿微米黑
        set(0, 'DefaultAxesFontName', 'WenQuanYi Micro Hei');
    catch
        set(0, 'DefaultTextFontName', 'Heiti TC');  % Mac系统备选黑体
        set(0, 'DefaultAxesFontName', 'Heiti TC');
    end
end

% 生成基础示例数据
rng(42);  % 设置随机种子，保证结果可重复
x = linspace(0, 10, 100);  % 生成0到10之间的100个均匀点
y = sin(x) + randn(1, 100)*0.2;  % 带噪声的正弦曲线
categories = {'A', 'B', 'C', 'D', 'E'};  % 类别标签
values = [23, 45, 56, 78, 32];  % 柱状图数据
bar_indices = 1:length(categories);  % 柱状图索引

% 1. 基础散点图
fprintf('显示第 1 种图表: 基础散点图...\n');
figure('Name', '基础散点图', 'Position', [100, 100, 800, 500]);
scatter(x, y, 50, 'b', 'filled', 'MarkerEdgeColor', 'k');  % 蓝色填充散点，黑色边缘
title('基础散点图');
xlabel('X轴');
ylabel('Y轴');
grid on;  % 显示网格
grid minor;  % 显示次要网格
set(gca, 'GridLineStyle', '--', 'GridAlpha', 0.7);  % 网格线样式为虚线，透明度0.7
box on;  % 显示坐标轴边框
pause(1);  % 暂停1秒，便于观察

% 2. 带趋势线的散点图
fprintf('显示第 2 种图表: 带趋势线的散点图...\n');
figure('Name', '带趋势线的散点图', 'Position', [100, 100, 800, 500]);
scatter(x, y, 50, 'b', 'filled', 'MarkerEdgeColor', 'k');
hold on;  % 保持当前图形，后续绘图叠加
p = polyfit(x, y, 1);  % 一阶多项式拟合（线性趋势）
y_trend = polyval(p, x);  % 计算趋势线数值
plot(x, y_trend, 'r--', 'LineWidth', 2);  % 红色虚线趋势线
title('带趋势线的散点图');
xlabel('X轴');
ylabel('Y轴');
legend('数据点', '趋势线', 'Location', 'best');  % 添加图例
grid on;
box on;
pause(1);

% 3. 分组散点图
fprintf('显示第 3 种图表: 分组散点图...\n');
figure('Name', '分组散点图', 'Position', [100, 100, 800, 500]);
groups = randi(3, 1, 100);  % 生成3个随机分组
colors = lines(3);  % 生成3种不同颜色
for i = 1:3
    idx = groups == i;  % 找到第i组的索引
    scatter(x(idx), y(idx), 50, colors(i,:), 'filled', 'MarkerEdgeColor', 'k');
    hold on;
end
title('分组散点图');
xlabel('X轴');
ylabel('Y轴');
legend('组1', '组2', '组3', 'Location', 'best');
grid on;
box on;
pause(1);

% 4. 气泡图
fprintf('显示第 4 种图表: 气泡图...\n');
figure('Name', '气泡图', 'Position', [100, 100, 800, 500]);
sizes = randi([10, 200], 1, 100);  % 气泡大小随机
scatter(x, y, sizes, sizes, 'filled', 'MarkerEdgeColor', 'k');  % 气泡大小和颜色都由sizes决定
colorbar;  % 显示颜色条
caxis([min(sizes), max(sizes)]);  % 设置颜色范围
title('气泡图');
xlabel('X轴');
ylabel('Y轴');
grid on;
box on;
pause(1);

% 5. 基础折线图
fprintf('显示第 5 种图表: 基础折线图...\n');
figure('Name', '基础折线图', 'Position', [100, 100, 800, 500]);
plot(x, sin(x), 'b-', 'LineWidth', 2);  % 蓝色实线，线宽2
title('基础折线图');
xlabel('X轴');
ylabel('Y轴');
grid on;
set(gca, 'GridAlpha', 0.3);  % 网格透明度0.3
box on;
pause(1);

% 6. 多系列折线图
fprintf('显示第 6 种图表: 多系列折线图...\n');
figure('Name', '多系列折线图', 'Position', [100, 100, 800, 500]);
% 同时绘制三条不同样式的曲线
plot(x, sin(x), 'b-', x, cos(x), 'r--', x, sin(x)+cos(x), 'g-.', 'LineWidth', 2);
title('多系列折线图');
xlabel('X轴');
ylabel('Y轴');
legend('sin(x)', 'cos(x)', 'sin(x)+cos(x)', 'Location', 'best');
grid on;
box on;
pause(1);

% 7. 带误差线的折线图
fprintf('显示第 7 种图表: 带误差线的折线图...\n');
figure('Name', '带误差线的折线图', 'Position', [100, 100, 800, 500]);
y_vals = sin(x);
error = rand(1, length(x)) * 0.2 + 0.1;  % 随机误差值
% 绘制带误差线的折线图
h = errorbar(x, y_vals, error, 'o-', 'MarkerFaceColor', 'b', 'LineWidth', 1.5, 'CapSize', 5);
set(h, 'Color', 'r');  % 设置线条颜色为红色
title('带误差线的折线图');
xlabel('X轴');
ylabel('Y轴');
grid on;
box on;
pause(1);

% 8. 阶梯折线图
fprintf('显示第 8 种图表: 阶梯折线图...\n');
figure('Name', '阶梯折线图', 'Position', [100, 100, 800, 500]);
step_data = cumsum(randn(1, 100));  % 随机游走数据
stairs(x, step_data, 'm-', 'LineWidth', 2);  % 品红色阶梯线
title('阶梯折线图');
xlabel('X轴');
ylabel('累积值');
grid on;
box on;
pause(1);

% 9. 基础柱状图
fprintf('显示第 9 种图表: 基础柱状图...\n');
figure('Name', '基础柱状图', 'Position', [100, 100, 800, 500]);
bar(bar_indices, values, 'FaceColor', [0.53, 0.81, 0.98]);  % 浅蓝色柱子
title('基础柱状图');
xlabel('类别');
ylabel('数值');
set(gca, 'XTick', bar_indices, 'XTickLabel', categories);  % 设置X轴标签
grid on;
set(gca, 'GridAlpha', 0.3, 'GridLineStyle', '--');
box on;
pause(1);

% 10. 水平柱状图
fprintf('显示第 10 种图表: 水平柱状图...\n');
figure('Name', '水平柱状图', 'Position', [100, 100, 800, 500]);
barh(bar_indices, values, 'FaceColor', [0.75, 0.91, 0.72]);  % 浅绿色水平柱子
title('水平柱状图');
xlabel('数值');
ylabel('类别');
set(gca, 'YTick', bar_indices, 'YTickLabel', categories);  % 设置Y轴标签
grid on;
set(gca, 'GridAlpha', 0.3, 'GridLineStyle', '--');
box on;
pause(1);

fprintf('所有10种图表显示完成！\n');

File: 资料/代码/Python绘图代码/基础散点图、带趋势线的散点图、分组散点图、气泡图、基础折线图、多系列折线图、带误差线的折线图、阶梯折线图、基础柱状图、水平柱状图.py
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
import matplotlib.colors as mcolors
import matplotlib

# 优化中文字体配置，使用更通用的字体方案
plt.rcParams["font.family"] = ["Microsoft YaHei", "WenQuanYi Micro Hei", "Heiti TC", "SimHei", "sans-serif"]
plt.rcParams["axes.unicode_minus"] = False  # 解决负号显示问题

# 检测并显示当前使用的字体
def check_font():
    try:
        # 尝试获取配置的字体
        font = matplotlib.font_manager.findfont(matplotlib.font_manager.FontProperties(family=plt.rcParams["font.family"]))
        print(f"当前使用的字体: {font}")
        return True
    except:
        print("警告: 未找到指定的中文字体，尝试备选方案...")
        # 简化的字体检测方案
        available_fonts = set()
        font_manager = matplotlib.font_manager.FontManager()
        for font in font_manager.ttflist:
            available_fonts.add(font.name)
        
        # 常见中文字体列表
        chinese_fonts = ["WenQuanYi Micro Hei", "Heiti TC", "SimHei", 
                        "Microsoft YaHei", "Arial Unicode MS", "SimSun", "NSimSun"]
        
        # 查找可用的中文字体
        for font in chinese_fonts:
            if font in available_fonts:
                plt.rcParams["font.family"] = [font, "sans-serif"]
                print(f"已自动切换到可用字体: {font}")
                return True
        
        # 如果没有找到中文字体，使用默认字体并警告
        print("警告: 系统中未找到可用的中文字体，中文可能无法正常显示。")
        print("建议安装WenQuanYi Micro Hei或SimHei等中文字体。")
        return False

# 设置图表风格
plt.style.use('seaborn-v0_8-dark-palette')
# 设置默认图像大小和分辨率
plt.rcParams["figure.figsize"] = (10, 6)
plt.rcParams["figure.dpi"] = 100

# 1. 基础散点图
def plot_scatter_basic():
    np.random.seed(42)
    x = np.random.randn(100) * 10
    y = 2 * x + np.random.randn(100) * 15 + 50
    
    plt.figure()
    plt.scatter(x, y, c='#2c7fb8', alpha=0.7, edgecolors='w', s=80, linewidth=1.5)
    plt.title('基础散点图', fontsize=16, pad=20)
    plt.xlabel('X轴数据', fontsize=12, labelpad=10)
    plt.ylabel('Y轴数据', fontsize=12, labelpad=10)
    plt.grid(True, linestyle='--', alpha=0.6)
    plt.tight_layout()
    plt.show()

# 2. 带趋势线的散点图
def plot_scatter_with_trend():
    np.random.seed(42)
    x = np.linspace(0, 10, 100)
    y = 3 * x + 15 + np.random.randn(100) * 5
    
    # 计算趋势线
    slope, intercept, r_value, p_value, std_err = stats.linregress(x, y)
    trend_line = slope * x + intercept
    
    plt.figure()
    plt.scatter(x, y, c='#e41a1c', alpha=0.6, edgecolors='w', s=60, linewidth=1, label='数据点')
    plt.plot(x, trend_line, 'b--', linewidth=2.5, 
             label=f'趋势线: y = {slope:.2f}x + {intercept:.2f}\nR² = {r_value**2:.4f}')
    plt.title('带趋势线的散点图', fontsize=16, pad=20)
    plt.xlabel('X轴', fontsize=12, labelpad=10)
    plt.ylabel('Y轴', fontsize=12, labelpad=10)
    plt.legend(loc='best', fontsize=10)
    plt.grid(True, linestyle='--', alpha=0.6)
    plt.tight_layout()
    plt.show()

# 3. 分组散点图
def plot_scatter_grouped():
    np.random.seed(42)
    groups = 4
    points_per_group = 50
    
    # 生成四组不同的数据
    x = [np.random.normal(i*3, 1.2, points_per_group) for i in range(groups)]
    y = [np.random.normal(i*2 + 5, 1.0, points_per_group) for i in range(groups)]
    labels = ['组A', '组B', '组C', '组D']
    colors = ['#a6cee3', '#1f78b4', '#b2df8a', '#33a02c']
    markers = ['o', 's', '^', 'D']
    
    plt.figure()
    for i in range(groups):
        plt.scatter(x[i], y[i], c=colors[i], label=labels[i], alpha=0.7, 
                   edgecolors='w', s=70, marker=markers[i], linewidth=1.5)
    
    plt.title('分组散点图', fontsize=16, pad=20)
    plt.xlabel('X轴指标', fontsize=12, labelpad=10)
    plt.ylabel('Y轴指标', fontsize=12, labelpad=10)
    plt.legend(loc='best', fontsize=10)
    plt.grid(True, linestyle='--', alpha=0.6)
    plt.tight_layout()
    plt.show()

# 4. 气泡图
def plot_bubble():
    np.random.seed(42)
    n = 50
    x = np.random.rand(n) * 100  # X坐标
    y = np.random.rand(n) * 100  # Y坐标
    sizes = np.random.rand(n) * 500 + 50  # 气泡大小
    colors = np.random.rand(n)  # 颜色值
    
    plt.figure()
    scatter = plt.scatter(x, y, s=sizes, c=colors, alpha=0.7, 
                         edgecolors='w', linewidth=1, cmap='viridis')
    
    # 添加颜色条
    cbar = plt.colorbar(scatter)
    cbar.set_label('重要程度', fontsize=12, labelpad=10)
    
    plt.title('气泡图', fontsize=16, pad=20)
    plt.xlabel('市场份额 (%)', fontsize=12, labelpad=10)
    plt.ylabel('增长率 (%)', fontsize=12, labelpad=10)
    plt.grid(True, linestyle='--', alpha=0.6)
    plt.tight_layout()
    plt.show()

# 5. 基础折线图
def plot_line_basic():
    np.random.seed(42)
    x = np.linspace(0, 12, 100)
    y = np.sin(x) * 10 + 50 + np.random.randn(100) * 2
    
    plt.figure()
    plt.plot(x, y, color='#984ea3', linewidth=2.5, linestyle='-', 
             marker='o', markersize=6, markerfacecolor='#ffff99', 
             markeredgecolor='#984ea3', markeredgewidth=1.5)
    
    plt.title('基础折线图', fontsize=16, pad=20)
    plt.xlabel('时间', fontsize=12, labelpad=10)
    plt.ylabel('数值', fontsize=12, labelpad=10)
    plt.grid(True, linestyle='--', alpha=0.6)
    plt.tight_layout()
    plt.show()

# 6. 多系列折线图
def plot_line_multi_series():
    np.random.seed(42)
    x = np.linspace(0, 10, 50)
    
    # 生成三个不同的系列
    y1 = np.sin(x) * 10 + 50 + np.random.randn(50) * 1.5
    y2 = np.cos(x) * 8 + 30 + np.random.randn(50) * 1.2
    y3 = (np.sin(x) + np.cos(x)) * 5 + 20 + np.random.randn(50) * 1.0
    
    plt.figure()
    plt.plot(x, y1, label='系列一', color='#e41a1c', linewidth=2, 
             marker='o', markersize=5, alpha=0.8)
    plt.plot(x, y2, label='系列二', color='#377eb8', linewidth=2, 
             linestyle='--', marker='s', markersize=5, alpha=0.8)
    plt.plot(x, y3, label='系列三', color='#4daf4a', linewidth=2, 
             linestyle='-.', marker='^', markersize=5, alpha=0.8)
    
    plt.title('多系列折线图', fontsize=16, pad=20)
    plt.xlabel('时间', fontsize=12, labelpad=10)
    plt.ylabel('数值', fontsize=12, labelpad=10)
    plt.legend(loc='best', fontsize=10)
    plt.grid(True, linestyle='--', alpha=0.6)
    plt.tight_layout()
    plt.show()

# 7. 带误差线的折线图
def plot_line_with_error():
    np.random.seed(42)
    x = np.arange(1, 13)  # 12个数据点，例如12个月
    y = np.sin(x/2) * 10 + 50 + np.random.randn(12) * 3
    y_err = np.random.rand(12) * 4 + 1  # 误差值
    
    plt.figure()
    plt.errorbar(x, y, yerr=y_err, fmt='-o', ecolor='#fb9a99', elinewidth=2,
                 capsize=5, capthick=2, color='#e41a1c', linewidth=2.5, 
                 markersize=8, markerfacecolor='#ffffff', markeredgewidth=2)
    
    plt.title('带误差线的折线图', fontsize=16, pad=20)
    plt.xlabel('月份', fontsize=12, labelpad=10)
    plt.ylabel('测量值', fontsize=12, labelpad=10)
    plt.xticks(x)  # 显示所有x刻度
    plt.grid(True, linestyle='--', alpha=0.6)
    plt.tight_layout()
    plt.show()

# 8. 阶梯折线图
def plot_step_line():
    np.random.seed(42)
    x = np.linspace(0, 10, 20)
    y = np.cumsum(np.random.randn(20) * 2 + 5)  # 累积和，模拟某种指标变化
    
    plt.figure()
    plt.step(x, y, where='mid', color='#ff7f00', linewidth=2.5, 
             marker='s', markersize=8, markerfacecolor='#ffffcc', 
             markeredgecolor='#ff7f00', markeredgewidth=2)
    
    plt.title('阶梯折线图', fontsize=16, pad=20)
    plt.xlabel('时间点', fontsize=12, labelpad=10)
    plt.ylabel('累计值', fontsize=12, labelpad=10)
    plt.grid(True, linestyle='--', alpha=0.6)
    plt.tight_layout()
    plt.show()

# 9. 基础柱状图
def plot_bar_basic():
    np.random.seed(42)
    categories = ['产品A', '产品B', '产品C', '产品D', '产品E', '产品F']
    values = np.random.randint(50, 200, len(categories))
    
    plt.figure()
    bars = plt.bar(categories, values, 
                   color=plt.cm.viridis(np.linspace(0, 0.9, len(categories))), 
                   edgecolor='w', linewidth=1.5)
    
    # 添加数据标签
    for bar in bars:
        height = bar.get_height()
        plt.text(bar.get_x() + bar.get_width()/2., height + 3,
                f'{height}', ha='center', va='bottom', fontsize=10, fontweight='bold')
    
    plt.title('基础柱状图', fontsize=16, pad=20)
    plt.xlabel('产品类别', fontsize=12, labelpad=10)
    plt.ylabel('销量', fontsize=12, labelpad=10)
    plt.grid(axis='y', linestyle='--', alpha=0.6)
    plt.tight_layout()
    plt.show()

# 10. 水平柱状图
def plot_bar_horizontal():
    np.random.seed(42)
    categories = ['地区A', '地区B', '地区C', '地区D', '地区E', '地区F']
    values = np.random.randint(100, 500, len(categories))
    
    plt.figure()
    bars = plt.barh(categories, values, 
                    color=plt.cm.plasma(np.linspace(0.2, 0.9, len(categories))), 
                    edgecolor='w', linewidth=1.5)
    
    # 添加数据标签
    for bar in bars:
        width = bar.get_width()
        plt.text(width + 10, bar.get_y() + bar.get_height()/2.,
                f'{width}', va='center', fontsize=10, fontweight='bold')
    
    plt.title('水平柱状图', fontsize=16, pad=20)
    plt.xlabel('销售额', fontsize=12, labelpad=10)
    plt.ylabel('地区', fontsize=12, labelpad=10)
    plt.grid(axis='x', linestyle='--', alpha=0.6)
    plt.tight_layout()
    plt.show()

# 运行所有图表
if __name__ == "__main__":
    # 检查字体配置，如果失败则提示用户
    font_available = check_font()
    if not font_available:
        print("注意：由于缺少中文字体，图表中的中文可能无法正常显示。")
    
    plot_scatter_basic()
    plot_scatter_with_trend()
    plot_scatter_grouped()
    plot_bubble()
    plot_line_basic()
    plot_line_multi_series()
    plot_line_with_error()
    plot_step_line()
    plot_bar_basic()
    plot_bar_horizontal()
